File: swift/Neetcode/TwoPointers/Medium/15.3Sum.swift
// Given an integer array nums, 
// return all the triplets [nums[i], nums[j], nums[k]] 
// such that i != j, i != k, and j != k, 
// and nums[i] + nums[j] + nums[k] == 0.

// Time: O(n^2), Space: O(1)
func threeSum(_ nums: [Int]) -> [[Int]] {
    let nums = nums.sorted()
    var result = [[Int]]()

    for i in 0..<(nums.count - 2) {
        if i > 0, nums[i - 1] == nums[i] { continue }

        var j = i + 1, k = nums.count - 1 

        while j < k {
            var sum = nums[i] + nums[j] + nums[k]

            if sum > 0 {
                k -= 1
            } else if sum < 0 {
                j += 1
            } else {
                result.append([nums[i], nums[j], nums[k]])
                var last = nums[j]

                while j < k, last == nums[j] {
                    j += 1
                }
            }
        }
    }

    return result
}

print(threeSum([-1,0,1,2,-1,-4]))
print(threeSum([0,1,1]))
print(threeSum([0,0,0]))

File: python/Neetcode/TwoPointers/Medium/15.3Sum.py
# Given an integer array nums, 
# return all the triplets [nums[i], nums[j], nums[k]] 
# such that i != j, i != k, and j != k, 
# and nums[i] + nums[j] + nums[k] == 0.

def threeSum(nums):
    nums.sort()
    result = []

    for i, number in enumerate(nums):
        if i != 0 and number == nums[i - 1]:
            continue

        left, right = i + 1, len(nums) - 1

        while left < right:
            total_sum = number + nums[left] + nums[right]

            if total_sum > 0:
                right -= 1
            elif total_sum < 0:
                left += 1
            else:
                result.append([number, nums[left], nums[right]])
                left += 1
                
                while nums[left] == nums[left - 1] and left < right:
                    left += 1

    return result

print(threeSum([-1,0,1,2,-1,-4]))
print(threeSum([0,1,1]))
print(threeSum([0,0,0]))